service/migration_manager.cc

```diff
@@ -1090,6 +1091,10 @@ future<schema_ptr> get_schema_definition(table_schema_version v, netw::messaging
                 // It should be fine since if we tried to register a view for which
                 // we don't know the base table, our registry is broken.
                 schema_ptr base_schema = db.find_schema(s->view_info()->base_id());
+                auto fixed_view = db::schema_tables::maybe_fix_legacy_secondary_index_mv_schema(get_local_storage_proxy().local_db(), view_ptr(s), base_schema, db::schema_tables::preserve_version::yes);
+                if (fixed_view) {
+                    s = fixed_view;
```

Comment:
I looked at the callers of `get_schema_definition()` and I don't really understand if and how this code will work. Schema pointers which describe the existing tables in a Scylla node have a special `_registry_entry` field which points to the schema registry entry. This entry (as far as I know) is only allocated for schema instances which were loaded from disk or announced via the migration manager, and is a `nullptr` otherwise. Here, we just build a temporary `schema_ptr` instance (`fixed_view`) and return it as if it was one of the registered schemas. The potential problem I see is that e.g. here:
https://github.com/scylladb/scylla/blob/dc7beec382535d9c55b0b6523ab5d2b9c94686d7/service/migration_manager.cc#L338-L340
we take the return value from `get_schema_definition` and dereference its registry entry, assuming that it exists. But does it, if we just generated this schema ourselves without registering it anywhere?

By the way - how was this change tested, and is there an issue I can read to get more context on why it's needed?

/cc @tgrabiec

Reply:
@psarna you are right, it smells bad. I will look into it.

Reply:
You are right, it is a problem, I should fix the view prior to registering it. I will make the changes.